src/test/librados/aio_cxx.cc

```diff
@@ -2466,3 +2466,281 @@ TEST(LibRadosAio, MultiReads) {
     ASSERT_EQ(0, memcmp(buf, bl.c_str(), sizeof(buf)));
   }
 }
+
+TEST(LibRadosAio, AioCancelBeforeSubmit)
+{
+  Rados cluster;
+  auto pool_prefix = fmt::format("{}_", ::testing::UnitTest::GetInstance()->current_test_info()->name());
+  std::string pool_name = get_temp_pool_name(pool_prefix);
+  ASSERT_EQ("", create_one_pool_pp(pool_name, cluster));
+
+  auto pool_cleanup = make_scope_guard([&] {
+        destroy_one_pool_pp(pool_name, cluster);
+      });
+
+  IoCtx ioctx;
+  cluster.ioctx_create(pool_name.c_str(), ioctx);
+
+  auto c = std::unique_ptr<AioCompletion>{Rados::aio_create_completion()};
+  ioctx.aio_cancel(c.get()); // don't crash
+}
+
+TEST(LibRadosAio, CompletionCancelBeforeSubmit)
+{
+  Rados cluster;
+  auto pool_prefix = fmt::format("{}_", ::testing::UnitTest::GetInstance()->current_test_info()->name());
+  std::string pool_name = get_temp_pool_name(pool_prefix);
+  ASSERT_EQ("", create_one_pool_pp(pool_name, cluster));
+
+  auto pool_cleanup = make_scope_guard([&] {
+        destroy_one_pool_pp(pool_name, cluster);
+      });
+
+  IoCtx ioctx;
+  cluster.ioctx_create(pool_name.c_str(), ioctx);
+
+  auto c = std::unique_ptr<AioCompletion>{Rados::aio_create_completion()};
+  c->cancel(); // don't crash
+}
+
+TEST(LibRadosAio, ReadOperationAioCancel)
+{
+  Rados cluster;
+  auto pool_prefix = fmt::format("{}_", ::testing::UnitTest::GetInstance()->current_test_info()->name());
+  std::string pool_name = get_temp_pool_name(pool_prefix);
+  ASSERT_EQ("", create_one_pool_pp(pool_name, cluster));
+
+  auto pool_cleanup = make_scope_guard([&] {
+        destroy_one_pool_pp(pool_name, cluster);
+      });
+
+  IoCtx ioctx;
+  cluster.ioctx_create(pool_name.c_str(), ioctx);
+
+  // cancellation tests are racy, so retry if completion beats the cancellation
+  int ret = 0;
+  int tries = 10;
+  do {
+    auto c = std::unique_ptr<AioCompletion>{Rados::aio_create_completion()};
+    ObjectReadOperation op;
+    op.assert_exists();
+    ioctx.aio_operate("nonexistent", c.get(), &op, nullptr);
+
+    ioctx.aio_cancel(c.get());
```

Comment:
thanks, good idea

i opted to remove the write tests because cancellation doesn't really care which aio_operate() overload we called to start it

i borrowed a fixture from test/librados/asio.cc to avoid recreating the pool each time

the fixture is also [value-parameterized](https://google.github.io/googletest/advanced.html#value-parameterized-tests) on the distinction between aio_cancel() vs cancel(). how does this look?
```
Note: Google Test filter = *Cancel*
[==========] Running 6 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 6 tests from LibRadosAio/Cancel
[ RUN      ] LibRadosAio/Cancel.BeforeSubmit/aio_cancel
[       OK ] LibRadosAio/Cancel.BeforeSubmit/aio_cancel (0 ms)
[ RUN      ] LibRadosAio/Cancel.BeforeSubmit/cancel
[       OK ] LibRadosAio/Cancel.BeforeSubmit/cancel (0 ms)
[ RUN      ] LibRadosAio/Cancel.BeforeComplete/aio_cancel
[       OK ] LibRadosAio/Cancel.BeforeComplete/aio_cancel (0 ms)
[ RUN      ] LibRadosAio/Cancel.BeforeComplete/cancel
[       OK ] LibRadosAio/Cancel.BeforeComplete/cancel (0 ms)
[ RUN      ] LibRadosAio/Cancel.AfterComplete/aio_cancel
[       OK ] LibRadosAio/Cancel.AfterComplete/aio_cancel (6 ms)
[ RUN      ] LibRadosAio/Cancel.AfterComplete/cancel
[       OK ] LibRadosAio/Cancel.AfterComplete/cancel (1 ms)
[----------] 6 tests from LibRadosAio/Cancel (7 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test suite ran. (2091 ms total)
[  PASSED  ] 6 tests.
```

Reply:
This looks great, just a few very minor nits which can be ignored.